### PR TITLE
pallet assets: optional auto-increment for the asset ID

### PR DESCRIPTION
**Pull request by muharem**
Introduce an optional auto-increment setup for the IDs of new assets.

### CODE REVIEW

File: substrate/frame/assets/src/lib.rs [hunk 1]
```diff
@@ -206,8 +207,37 @@ pub trait AssetsCallback<AssetId, AccountId> {
...
+	fn destroyed(id: &AssetId) -> Result<(), ()> {
+		for_tuples!( #( Tuple::destroyed(id)?; )* );
+		Ok(())
+	}
+}
+
+/// Auto-increment the [`NextAssetId`] when an asset is created.
+///
+/// This has not effect if the [`NextAssetId`] value is not present.
+pub struct AutoIncrAssetId<T, I = ()>(PhantomData<(T, I)>);
```
Comment (joepetrowski) [inline substrate/frame/assets/src/lib.rs:226]: I think I'd either call it `AutoIncrementAssetId` or `AutoIncAssetId`

File: substrate/frame/assets/src/lib.rs [hunk 2]
```diff
@@ -622,6 +666,8 @@ pub mod pallet {
 		NotFrozen,
 		/// Callback action resulted in error
 		CallbackFailed,
+		/// The asset id must be equal to the [`NextAssetId`].
```
Comment (joepetrowski) [inline substrate/frame/assets/src/lib.rs:669]:
```suggestion
		/// The asset ID must be equal to the [`NextAssetId`].
```

**Review by joepetrowski — APPROVED** [review]

**Comment by muharem** [comment]
@bkchr @ggwpez  can you help with review here, it would be helpful to get this into the next sdk release

File: substrate/frame/assets/src/migration.rs [hunk 3]
```diff
@@ -22,6 +22,26 @@ use log;
...
+	pub struct SetNextAssetId<ID, T: Config<I>, I: 'static = ()>(
+		core::marker::PhantomData<(ID, T, I)>,
+	);
+	impl<ID, T: Config<I>, I: 'static> OnRuntimeUpgrade for SetNextAssetId<ID, T, I>
+	where
+		T::AssetId: Incrementable,
+		ID: Get<T::AssetId>,
+	{
+		fn on_runtime_upgrade() -> frame_support::weights::Weight {
+			NextAssetId::<T, I>::put(ID::get());
```
Comment (bkchr) [inline substrate/frame/assets/src/migration.rs:39]:
```suggestion
			if !NextAssetId::<T, I>::exists() {
				NextAssetId::<T, I>::put(ID::get());
			}
```

File: substrate/frame/assets/src/migration.rs [hunk 4]
```diff
@@ -22,6 +22,26 @@ use log;
 #[cfg(feature = "try-runtime")]
 use sp_runtime::TryRuntimeError;
 
+pub mod next_asset_id {
+	use super::*;
+	use sp_core::Get;
+
+	/// Set [`NextAssetId`] to the value of `ID`.
```
Comment (bkchr) [inline substrate/frame/assets/src/migration.rs:29]:
```suggestion
	/// Set [`NextAssetId`] to the value of `ID` iff [`NextAssetId`] does not exist yet.
```

File: substrate/frame/assets/src/lib.rs [hunk 5]
```diff
@@ -415,6 +450,15 @@ pub mod pallet {
...
 	>;
 
+	/// The asset ID to be used for the next asset creation, if any present. Otherwise, it has no
+	/// effect.
+	///
+	/// The next asset ID can be set using the
+	/// [SetNextAssetId](`migration::next_asset_id::SetNextAssetId`) migration. For the auto-
+	/// incremented model, the [`crate::AutoIncAssetId`] callback can be used.
+	#[pallet::storage]
+	pub type NextAssetId<T: Config<I>, I: 'static = ()> = StorageValue<_, T::AssetId, OptionQuery>;
```
Comment (bkchr) [inline substrate/frame/assets/src/lib.rs:460]: There is no way to set this at genesis. Can you please add it to the `GenesisConfig`?

**Review by bkchr — APPROVED** [review]
The implementation is a little bit complicated ;) I mean it doesn't break anything or at least not that much. 

Left some things to improve. A followup would be nice that improves the docs on how to use this.

File: substrate/frame/assets/src/lib.rs (hunk 5 above)
Comment (muharem) [reply substrate/frame/assets/src/lib.rs:463]: I did not touch it to make this non breaking change for most of the users. But no strong opinion. 
You can see it now in `GenesisConfig`.

**Comment by bkchr** [comment]
> Should we wait for an audit or we merge this?

I think we can go ahead.

**Comment by DrW3RK** [comment]
Hey @muharem, What was the motivation behind this PR? 

We need to update the docs on Asset creation and would like to know why users cannot claim asset IDs that are unclaimed and are smaller than the nextAssetID.

**Comment by muharem** [comment]
@DrW3RK with auto-increment it is easy to make asset ids not reusable. without auto-increment you can use any ids which is not allocated for any asset at the moment.

**Pull request merged**